tests/unit/steps/test_plan.py

```diff
@@ -1052,6 +1051,107 @@ def test_get_pre_upgrade_steps(cli_backup, cli_args, model):
     assert pre_upgrade_steps == expected_steps
 
 
+@patch("cou.steps.plan._get_purge_data_steps")
+@patch("cou.steps.plan._get_archive_data_steps")
+@pytest.mark.parametrize(
+    "case,archive,purge",
+    [
+        ("Empty steps", False, False),
+        ("Archive", True, False),
+        ("Purge", False, True),
+        ("Archive and Purge", False, True),
+    ],
+)
+def test_get_pre_clean_data_steps(
+    mock_get_archive_data_steps,
+    mock_get_purge_data_steps,
+    case,
+    purge,
+    archive,
+    cli_args,
+    model,
+):
+    cli_args.purge = purge
+    cli_args.archive = archive
+    mock_analysis_result = MagicMock(spec=Analysis)()
+    mock_analysis_result.model = model
+
+    mock_get_archive_data_steps.return_value = ["archive_step"]
+    mock_get_purge_data_steps.return_value = ["purge_step"]
+    steps = cou_plan._get_pre_clean_data_steps(mock_analysis_result, cli_args)
+
+    expected_steps = []
+
+    if archive:
+        mock_get_archive_data_steps.assert_called_with(mock_analysis_result, cli_args)
+        expected_steps.append("archive_step")
+    if purge:
+        mock_get_purge_data_steps.assert_called_with(mock_analysis_result, cli_args)
+        expected_steps.append("purge_step")
+    assert expected_steps == steps
+
+
+def test_get_archive_data_steps(cli_args, model):
+    cli_args.archive_batch_size = 2000
+    mock_analysis_result = MagicMock(spec=Analysis)()
+    mock_analysis_result.model = model
+
+    steps = cou_plan._get_archive_data_steps(mock_analysis_result, cli_args)
+    assert steps == [
+        PreUpgradeStep(
+            description="Archive old database data on nova-cloud-controller",
+            coro=archive(mock_analysis_result.model, batch_size=2000),
+        )
+    ]
+
+
+@pytest.mark.parametrize(
+    "case,charm_exists,action_exists,purge_before",
+    [
+        ("No action exists", True, False, None),
+        ("No application exists", False, True, None),
+        ("Purge all", True, True, None),
+        ("Purge before", True, True, "2000-01-02"),
+    ],
+)
+def test_get_purge_data_steps(case, charm_exists, action_exists, purge_before, cli_args, model):
+    cli_args.purge = True
+    cli_args.purge_before = purge_before
+    mock_analysis_result = MagicMock(spec=Analysis)()
+    mock_analysis_result.model = model
+
+    if charm_exists:
```

Comment:
In this case because `charm_exists`, `action_exists`, and `purge_before` all influence the result. Separate them is just like remove the `pytest.mark.parametrize` but all the code need to be copy.

Reply:
For this case because the purpose is to check the return steps exists only if `charm_exists` and `action_exists`. The suggestion is like remove the pytest.mark.parametrize and copy the code everywhere.

Reply:
I still think that there is too much logic in this test and the maintainability won't be easy.

Maybe split the test into what is expected to be empty and what is not. It's easier to understand `expected_steps == []` Then trying to figure out the parametrization options that will result empty.

Reply:
Update in the newest commit.